Refactor PromQL function handling

saswatamcode opened the pull request:
This PR modifies physical plan to ensure that the engine is able to support multi-argument PromQL functions. It categorizes PromQL functions into different categories and changes query plan tree accordingly, using a function_selector operator.

With this, the engine can now support single instant vector, single range vector, single scalar input functions as well as combined instant vector and scalar input functions. WIll update this PR to also handle combined range vector and scalar functions and variadic ones.

Also while refactoring function logic, this adds support for `parser.MatrixSelector` queries like `http_requests_total[1m]`.

Fixes #23 and #39.

fpetkovski reviewed (commented):
Thanks for this PR, I think the direction needs a bit of adjustment, but there is already a lot of code we can reuse. I would make the following suggestions:

First, I would make function to a standalone operator similar to how unary was implemented here: https://github.com/thanos-community/promql-engine/pull/56, it probably does not belong in the scan package. I would also suggest that each function does its own argument parsing instead of having this logic bundled in one central place, similar to what was done in https://github.com/thanos-community/promql-engine/pull/46 and https://github.com/thanos-community/promql-engine/pull/60

I would also leave the special handling of call + matrix selector as it is, we can resolve that case later once we have proper support for arbitrary functions. Let's rather start small and add support for clamp_min and clamp_max, then we can thing about more complex cases.

saswatamcode commented:
Thanks!

> I would make function to a standalone operator similar to how unary was implemented here: https://github.com/thanos-community/promql-engine/pull/56, it probably does not belong in the scan package. I would also suggest that each function does its own argument parsing instead of having this logic bundled in one central place, similar to what was done in https://github.com/thanos-community/promql-engine/pull/46 and https://github.com/thanos-community/promql-engine/pull/60

Yup I was thinking the same after looking at unary PR! Will do.
For the argument parsing, the reason why I was doing it in centralized way, was that `FunctionCall` only has `promql.Point` args, but it is difficult to deduce type from them. Will try to look on ways of handling that better! 🙂 

> I would also leave the special handling of call + matrix selector as it is, we can resolve that case later once we have proper support for arbitrary functions. Let's rather start small and add support for clamp_min and clamp_max, then we can think about more complex cases.

I think call + matrix selector already works here. I'll try to clean it up a bit. This doesn't add support for multi-arg range vector functions or variadic ones. Will iterate on those in future PRs!

fpetkovski commented on physicalplan/scan/matrix_selector.go:
```diff
@@ -27,10 +27,15 @@ type matrixScanner struct {
-	funcExpr *parser.Call
-	call     FunctionCall
 	storage  *seriesSelector
+	in       *Injectable
```
Hm I don't understand why this change is necessary. We should always be able to inject a function in a matrix selector. Otherwise the return type will not match.

saswatamcode replied:
So earlier when we only handled `parser.MatrixSelector` via `parser.Call`, we could always inject the function directly. But now that it might be called a separate operator using queries like `http_requests_total[30s]`, it might not always need a call.

Also, I wanted to create a mechanism to inject some modifications while creating a new operator recursively, and this seemed like a way to achieve that. This can be used in the future when we might want to create different modified operators based on queries. Wdyt? 🙂

fpetkovski replied:
For a query like `http_requests_total[30s]`, the return type is a really a matrix, not a vector. So I don't think that query will fit into the model that we have now, and I suspect we might need a completely different operator for that case.

Let's maybe try to tackle that problem separately, and use this PR to enable function calls with an arbitrary number of parameters.

saswatamcode replied:
Yup, I'll add it as a TODO and remove.

saswatamcode replied:
Have modified this to only have `call` for now. Added TODO

fpetkovski commented on physicalplan/scan/literal_selector.go:
```diff
@@ -45,30 +41,7 @@ func NewNumberLiteralSelector(pool *model.VectorPool, mint, maxt time.Time, step
-		callName:    f.Name,
-	}, nil
-}
-
```
👍

fpetkovski commented on physicalplan/scan/matrix_selector.go:
```diff
@@ -127,19 +127,44 @@ func (o *matrixSelector) Next(ctx context.Context) ([]model.StepVector, error) {
+					vectors[currStep].Samples = append(vectors[currStep].Samples, result.V)
+					vectors[currStep].SampleIDs = append(vectors[currStep].SampleIDs, series.signature)
+				}
+			} else {
```
Let's leave this out for now and revert this change. We can fall back to the old engine in this case and maybe open an issue to tackle it in the future.

saswatamcode replied:
Yup, issue is already open #39. I'll add a TODO for now. This is also why I added the injectable type.

fpetkovski commented on physicalplan/scan/matrix_selector.go:
```diff
@@ -170,9 +195,6 @@ func (o *matrixSelector) loadSeries(ctx context.Context) error {
 		o.series = make([]labels.Labels, len(series))
 		for i, s := range series {
 			lbls := s.Labels()
-			if o.funcExpr.Func.Name != "last_over_time" {
```
Is this safe to do?

yeya24 replied:
This seems to be moved to https://github.com/thanos-community/promql-engine/pull/62/files#diff-c5565c814b1f8018448d49f7dc602d00b7cf0225383932129b1d5ba967bfafa0R257

fpetkovski commented on physicalplan/unary/unary.go:
```diff
@@ -27,16 +30,22 @@ type unaryNegation struct {
 	return u, nil
 }
 
+func (o *unaryNegation) Explain() (me string, next []model.VectorOperator) {
```
Interesting that this wasn't added. Should we have a separate `Explainable` interface and make `Explain` return 
```
(me string, next []Explainable) 
```
so that we don't miss such cases in the future?

We can also add the method on `model.VectorOperator`, but having a separate interface will help with reusing the same mechanism for visualizing the logical plan.

cc @bwplotka

bwplotka replied:
👍🏽

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -1,101 +1,279 @@
+	Points       []promql.Point
+	StepTime     time.Time
+	SelectRange  time.Duration
+	ScalarPoints []promql.Point
```
Can we make this:
```suggestion
	ScalarPoints []float64
```

I don't think we can specify points in PromQL functions.

saswatamcode replied:
So this is basically data points we get from the scalar arguments of a function like `clamp(instant vector, scalar, scalar)`. But yes this can be float64. I'll update!

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -1,101 +1,279 @@
+
+func NewFunctionSelector(pool *model.VectorPool, funcExpr *parser.Call, mint, maxt time.Time, step time.Duration, stepsBatch int, call FunctionCall, majorArgIndex int, majorArg model.VectorOperator, scalarArgs ...model.VectorOperator) model.VectorOperator {
+	return &functionSelector{
+		vectorPool:    pool,
```
I think this operator can make all changes in-place, like for the unary operator. So we should not need a new pool here, and we can return the `next` pool in the `GetPool()` function.

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -378,6 +556,32 @@ func dropMetricName(l labels.Labels) labels.Labels {
 	return labels.NewBuilder(l).Del(labels.MetricName).Labels(nil)
 }
 
+func clamp(points []promql.Point, scalars []promql.Point) (float64, bool) {
```
```suggestion
func clamp(points []promql.Point, scalars []float64) (float64, bool) {
```

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -410,13 +614,46 @@ func maybeSample(lbls labels.Labels, stepTime time.Time, points []promql.Point,
 	return f()
 }
 
+func ignoreEmptyScalar(points, scalars []promql.Point, f func() promql.Sample) promql.Sample {
```
Why do we need this function? Scalars should always be available and the syntax will be validated by the PromQL parser.

fpetkovski commented on physicalplan/plan.go:
```diff
@@ -121,18 +177,19 @@ func newOperator(expr parser.Expr, storage storage.Queryable, mint time.Time, ma
 		return newCancellableOperator(e.Expr, storage, mint, maxt, step, lookbackDelta)
 
 	case *parser.StringLiteral:
+		// TODO(saswatamcode): Ensure Explain does not panic.
```
Let's maybe remove handling of `StringLiteral` to resolve the panic. We should stabilize what we have now so that we can run the engine in actual environments.

saswatamcode replied:
Will remove!

saswatamcode replied:
Have removed the string literal support that we had previously for now. This would likely need some new operator!

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -1,101 +1,279 @@
+type functionSelector struct {
+	vectorPool    *model.VectorPool
+	funcExpr      *parser.Call
+	mint          int64
+	maxt          int64
+	step          int64
+	currentStep   int64
+	stepsBatch    int
```
Do we need these parameters? Functions don't usually select directly from storage.

saswatamcode replied:
Will remove!

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -1,101 +1,279 @@
+		}
+
+		switch o.funcExpr.Args[o.majorArgIndex].Type() {
+		case parser.ValueTypeVector, parser.ValueTypeScalar:
```
Hm, keep in mind this decision is not made during plan execution, but is really possible to make it at plan compilation because the syntax tree won't change. So Maybe we can precalculate things in the operator constructor.

The next function should be as tight as possible, and basically do
* call next on dowstream
* apply function
* sent result

saswatamcode replied:
I'll try to modify this!

saswatamcode replied:
Have modified this to now parse in the operator constructor itself!

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -1,101 +1,279 @@
+	stepsBatch    int
+	series        []labels.Labels
+	once          sync.Once
+	majorArgIndex int
```
Do we need to know the index?

saswatamcode replied:
Yes, as we use that to categorize the type of function. While the major arg is usually at [0] of PromQL function arguments, there are cases like `histogram_quantile(scalar, instant vector)` where it is at [1] instead. So need to look out for those cases.

fpetkovski commented on physicalplan/function/function.go:
```diff
@@ -1,101 +1,279 @@
+	series        []labels.Labels
+	once          sync.Once
+	majorArgIndex int
+	majorArg      model.VectorOperator
```
```suggestion
	next      model.VectorOperator
```

fpetkovski reviewed (commented):
This looks much much better now. We might want to add a benchmark for clamp_max to make sure we are not actually worse than the previous engine for some reason.

yeya24 commented on physicalplan/unary/unary.go:
```diff
@@ -27,16 +30,22 @@ type unaryNegation struct {
 	u := &unaryNegation{
 		next:       next,
 		stepsBatch: stepsBatch,
+		op:         op,
```
Umm why do we need OP? It is unary negation so op is `-`. For `+` we don't have an additional operator to handle that, same does promql.

saswatamcode replied:
You're right! Will remove the `op`.

saswatamcode commented:
Have rebased the PR and addressed most of the feedback! This also removes the string literal support that was there previously, as it was a bit hacky. String literal and range vector would likely need some changes in model, so plan to do those in the future. Stabilizing what is already there! 🙂

saswatamcode commented on physicalplan/function/function.go:
```diff
@@ -1,247 +1,479 @@
-type FunctionCall func(labels labels.Labels, points []promql.Point, stepTime int64, selectRange int64) promql.Sample
+type FunctionCall func(f FunctionArgs) promql.Sample
 
 var Funcs = map[string]FunctionCall{
```
This map can probably be split into different categories of function, so that the arguments align with PromQL. This would make processing them easier.

saswatamcode commented on physicalplan/function/function.go:
```diff
@@ -1,247 +1,479 @@
+
+	out := o.next.GetPool().GetVectorBatch()
+	for _, vector := range input {
+		step := o.next.GetPool().GetStepVector(vector.T)
```
We can modify step vector in-place here.

saswatamcode reviewed (commented):
Some comments from 1:1 with @fpetkovski. 🙂

saswatamcode commented:
Simplified it and added support for scalars which can select from storage. This still needs some work to add support for `scalar()`, and to streamline operator tree construction for different types of functions!

bwplotka commented on engine/engine_test.go:
```diff
@@ -765,6 +765,60 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
+			http_requests_total{pod="nginx-2"} 1+2x18`,
+			query: `clamp(rate(http_requests_total[30s]), 10 - 5, 10)`,
+		},
+		// {
```
?

saswatamcode replied:
Yes I need to implement `scalar()` as mentioned here https://github.com/thanos-community/promql-engine/pull/62#issuecomment-1280002880 🙂

saswatamcode replied:
This is done in https://github.com/thanos-community/promql-engine/pull/62/commits/95ece7c570466cc589079d31224a11826bc5fd64

bwplotka reviewed (commented):
Generally looks good (:

saswatamcode commented on physicalplan/scan/matrix_selector.go:
```diff
@@ -83,9 +81,6 @@ func NewMatrixSelector(
 func (o *matrixSelector) Explain() (me string, next []model.VectorOperator) {
 	r := time.Duration(o.selectRange) * time.Millisecond
-	if o.call != nil {
-		return fmt.Sprintf("[*matrixSelector] %v({%v}[%s] %v mod %v)", o.funcExpr.Func.Name, o.storage.matchers, r, o.shard, o.numShards), nil
```
This goes through the function operator, so explain statements would be added there.

saswatamcode commented on physicalplan/plan.go:
```diff
@@ -86,46 +87,58 @@ func newOperator(expr parser.Expr, storage storage.Queryable, opts *query.Option
-				operators = append(operators, operator)
-			}
+	case *parser.Call:
+		// We can classify PromQL functions into the following categories,
```
Will add

fpetkovski commented on execution/scan/matrix_selector.go:
```diff
@@ -123,14 +118,32 @@ func (o *matrixSelector) Next(ctx context.Context) ([]model.StepVector, error) {
+			} else {
+				// TODO(saswatamcode): Range vector result might need new operator.
+				// https://github.com/thanos-community/promql-engine/issues/39
+				return nil, parse.ErrNotImplemented
```
This error needs to be returned when the operator is created.

saswatamcode replied:
Moved the error

saswatamcode commented on physicalplan/function/function.go:
```diff
@@ -1,247 +1,479 @@
 )
 
+// functionSelector returns []model.StepVector after processing input with desired function.
+type functionSelector struct {
```
Renamed!

fpetkovski commented on engine/engine_test.go:
```diff
@@ -786,6 +786,78 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
+		//  	http_requests_total{pod="nginx-1"} 1+1x15
+		//  	http_requests_total{pod="nginx-2"} 1+2x18`,
+		// 	query: `scalar(http_requests_total)`,
+		// },
```
Maybe we can add a new test case which does not use `testutil`. Possibly a dedicated test for error cases. We can create a new issue/PR for this.

saswatamcode replied:
Created issue #88! Not implementing in this PR as it might be complicated, will follow up with a future one. Keeping TODO comment. 🙂

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+		nextOps:   nextOps,
+		call:      call,
+		funcExpr:  funcExpr,
+		nextIndex: 0,
```
```suggestion
		vectorIndex: 0,
```

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+
+	for i := range funcExpr.Args {
+		if funcExpr.Args[i].Type() == parser.ValueTypeMatrix || funcExpr.Args[i].Type() == parser.ValueTypeVector {
+			f.nextIndex = i
```
```suggestion
			f.nextIndex = i
			break
```

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+	case parser.ValueTypeMatrix:
+		f.nextFunc = f.rangeVectorNext
+	default:
+		f.nextFunc = f.defaultNext
```
Let's throw the exception here so that we can catch it and fall back.

saswatamcode replied:
Added error.

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+	}
+
+	// Call next on all scalar inputs.
+	scalarInput := make([][]model.StepVector, len(o.nextOps)-1)
```
Can we remove allocs by using a shared slice scoped to the operator?

saswatamcode replied:
Have added slice to struct! 🙂

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+		}
+	}
+
+	scalarPoints := make([]float64, len(scalarInput))
```
Can be potentially be shared across `Next` invocations

saswatamcode replied:
Have added slice to struct! 🙂

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+	}
+
+	scalarPoints := make([]float64, len(scalarInput))
+	out := o.nextOps[o.nextIndex].GetPool().GetVectorBatch()
```
Let's see if we can modify the input batch in place.

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+
+	// We inject function call in tree, so only need to relay input from here.
+	return input, nil
+}
```
Let's revert the Matrix selector nesting under Call so that we can remove the indirection here

saswatamcode replied:
Reverted.

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+	o.once.Do(func() {
+		if o.funcExpr.Func.Name == "vector" {
+			o.series = []labels.Labels{labels.New()}
+			o.nextOps[o.nextIndex].GetPool().SetStepSize(len(o.series))
```
```suggestion
```

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+
+		if o.funcExpr.Func.Name == "scalar" {
+			o.series = []labels.Labels{}
+			o.nextOps[o.nextIndex].GetPool().SetStepSize(len(o.series))
```
```suggestion
```

fpetkovski commented on execution/function/function.go:
```diff
@@ -0,0 +1,833 @@
+			o.series[i] = lbls
+		}
+
+		o.nextOps[o.nextIndex].GetPool().SetStepSize(len(o.series))
```
```suggestion
```

fpetkovski commented on execution/function/function.go:
```diff
@@ -1,247 +1,482 @@
+			// Scalar expressions can select from storage as well.
+			if len(o.funcExpr.Func.ArgTypes) > 1 && len(o.scalarInput) > 0 {
+				for l := range o.scalarInput {
+					if len(o.scalarInput[l]) >= 1 {
```
What happens if this condition is not met?

saswatamcode replied:
I thought of adding it as a safeguard against empty input. But yup, a test case is necessary here I think. Might need to handle this differently. Empty series cases could also be a part of #88 I think. 🙂 
Will update this one

saswatamcode replied:
I've added handling for this and verified empty cases manually. But a lot of it returns as NaN in certain cases. So will add test cases for this in #88

fpetkovski commented on execution/function/function.go:
```diff
@@ -1,247 +1,482 @@
+			}
+
+			o.scalarInput[s] = si
+			defer o.nextOps[i].GetPool().PutVectors(o.scalarInput[s])
```
I think this will miss the recycling of the vector operator samples

saswatamcode replied:
Have added defer `PutStepVector` statements to release samples. Both for `scalarInput` as well as `input`

fpetkovski commented on execution/function/function.go:
```diff
@@ -1,247 +1,482 @@
+				lbls = DropMetricName(s)
+			}
+
+			sort.Sort(lbls)
```
I think labels will always be sorted, so maybe we can remove this to save some CPU cycles?

fpetkovski reviewed (approved):
Great work on this, looks good to me 👍 

Let's give other reviewers a bit more time and we can merge.

yeya24 commented on execution/function/function.go:
```diff
@@ -1,247 +1,474 @@
+
+			vector.Samples = vector.Samples[:1]
+			vector.SampleIDs = vector.SampleIDs[:1]
+			vector.Samples[0] = math.NaN()
```
Why do we set this to nan?

saswatamcode replied:
`scalar()` will return NaN if the given input vector has more than one sample as mentioned here: https://prometheus.io/docs/prometheus/latest/querying/functions/#scalar.

yeya24 replied:
Thanks for the explanation.

yeya24 commented on execution/function/function.go:
```diff
@@ -1,247 +1,474 @@
+		series, loadErr := o.nextOps[o.vectorIndex].Series(ctx)
+		if loadErr != nil {
+			err = loadErr
+		}
```
Do we need to return in this case?

saswatamcode replied:
Will add

yeya24 commented on engine/engine_test.go:
```diff
@@ -786,6 +786,78 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
+			load: `load 30s
+			http_requests_total{pod="nginx-1"} 1+1x15
+			http_requests_total{pod="nginx-2"} 1+2x18`,
+			query: `clamp_min(http_requests_total, scalar(max(http_requests_total)) + 10)`,
```
Can we update the readme with the new benchmark result?

saswatamcode replied:
Sure will do!

saswatamcode replied:
Updated the single-core one. For multi-core I don't have an 8-core machine 😅

saswatamcode commented on execution/execution.go:
```diff
@@ -81,50 +82,74 @@ func newOperator(expr parser.Expr, storage *engstore.SelectorPool, opts *query.O
+		// - No inputs (only pi() and time()). TODO(saswatamcode)
+		// - Only scalar input (only vector()).
+		// - Variadics (all time based functions like month() + round() and label_join()). TODO(saswatamcode)
+		// Tracked in https://github.com/thanos-community/promql-engine/issues/23
```
Sure will remove!

yeya24 commented on README.md:
```diff
@@ -89,38 +89,68 @@ Note that memory usage is higher when executing a query with parallelism greater
+RangeQuery/clamp_min-8                                   84.4k ± 0%    802.2k ± 0%  +850.57%  (p=0.000 n=10+9)
+RangeQuery/complex_func_query-8                           100k ± 0%      817k ± 0%  +717.42%  (p=0.000 n=10+9)
+RangeQuery/func_within_func_query-8                      93.7k ± 0%    818.3k ± 0%  +772.88%  (p=0.000 n=10+10)
+RangeQuery/aggr_within_func_query-8                      93.7k ± 0%    818.3k ± 0%  +772.81%  (p=0.000 n=10+9)
```
Probably let's have an issue to track the allocation regression. 10x seems not normal, maybe we could have some kind of pooling.

fpetkovski replied:
Agree, I also noticed this which is strange. But total bytes used seems to be in line which is why I was not concerned too much.

saswatamcode replied:
Yup! Will create an issue. There isn't a memory/cpu regression but allocs seem abnormally high

yeya24 commented on execution/parse/errors.go:
```diff
@@ -17,3 +17,5 @@ func UnsupportedOperationErr(op parser.ItemType) error {
 	return errors.Wrap(ErrNotSupportedExpr, msg)
 }
+
+var ErrNotImplemented = errors.New("expression not implemented")
```
Thanks for the explanation.

yeya24 reviewed (approved):
Thanks for the great work!
